plugin/nsid/nsid.go

```diff
@@ -0,0 +1,97 @@
+// Package nsid implements NSID protocol
+package nsid
+
+import (
+	"encoding/hex"
+
+	"github.com/coredns/coredns/plugin"
+
+	"github.com/miekg/dns"
+	"golang.org/x/net/context"
+)
+
+// Nsid plugin
+type Nsid struct {
+	Next plugin.Handler
+	Data string
+}
+
+// ResponseWriter is a response writer that adds NSID response
+type ResponseWriter struct {
+	dns.ResponseWriter
+	Data string
+}
+
+// ServeDNS implements the plugin.Handler interface.
+func (n Nsid) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	if option := r.IsEdns0(); option != nil {
+		for _, o := range option.Option {
+			if _, ok := o.(*dns.EDNS0_NSID); ok {
+				nw := &ResponseWriter{ResponseWriter: w, Data: n.Data}
+				return plugin.NextOrFailure(n.Name(), n.Next, ctx, nw, r)
+
+			}
+		}
+	}
+	return plugin.NextOrFailure(n.Name(), n.Next, ctx, w, r)
+}
+
+// WriteMsg implements the dns.ResponseWriter interface.
+func (w *ResponseWriter) WriteMsg(res *dns.Msg) error {
+	r1 := new(dns.Msg)
+
+	// Make a copy of Msg and update NSID
```

Comment:
Why do you copy, you should be able to get away by just adding the NSID?

Reply:
Thanks @miekg.

Initially I modified `e.Nsid` directly, but incorrectly not converting buffer to hex in `e.Nsid`. That results in `overflow packing` error. I though the error was caused by direct modification of `e.Nsid` and copy is needed.

Later on I fixed the hex issue, but forgot to give the direct modification a try. So the usage of copy remains.

Just tried again with direct modification and it works now. Thanks!